f3/granite.go

```diff
@@ -609,68 +799,117 @@ type chainPower struct {
+	senderPower, _ := q.powerTable.Get(sender)
+	sigCopy := make([]byte, len(signature))
+	copy(sigCopy, signature)
 	if ok {
 		// Don't double-count the same chain head for a single participant.
-		for _, old := range fromSender.heads {
-			if head == old {
-				return
-			}
+		if _, ok := fromSender.heads[head]; ok {
+			return
 		}
-		fromSender.heads = append(fromSender.heads, head)
+		fromSender.heads[head] = sigCopy
 	} else {
 		// Add sender's power to total the first time a value is received from them.
-		senderPower, _ := q.powerTable.Get(sender)
 		q.sendersTotalPower.Add(q.sendersTotalPower, senderPower)
-		fromSender = senderSent{[]TipSetID{head}, senderPower}
+		fromSender = senderSent{
+			heads: map[TipSetID][]byte{
+				head: sigCopy,
+			},
+			power: senderPower,
+		}
 	}
 	q.received[sender] = fromSender
 
-	power, _ := q.powerTable.Get(sender)
-	candidate := chainPower{
+	candidate := chainSupport{
 		chain:           value,
-		power:           power,
+		power:           senderPower,
+		signers:         make(map[ActorID]struct{}),
 		hasStrongQuorum: false,
 		hasWeakQuorum:   false,
 	}
-	if found, ok := q.chainPower[head]; ok {
+	if found, ok := q.chainSupport[head]; ok {
 		candidate.power.Add(candidate.power, found.power)
+		candidate.signers = found.signers
 	}
+	candidate.signers[sender] = struct{}{}
 
 	candidate.hasStrongQuorum = hasStrongQuorum(candidate.power, q.powerTable.Total)
 	candidate.hasWeakQuorum = hasWeakQuorum(candidate.power, q.powerTable.Total)
 
-	q.chainPower[head] = candidate
+	if !value.IsZero() && justification.Payload.Step == PREPARE_PHASE { //only committed roundStates need to store justifications
+		q.justifiedMessages[value.Head().CID] = justification
+	}
+	q.chainSupport[head] = candidate
 }
 
 // Checks whether a value has been received before.
 func (q *quorumState) HasReceived(value ECChain) bool {
-	_, ok := q.chainPower[value.HeadCIDOrZero()]
+	_, ok := q.chainSupport[value.HeadCIDOrZero()]
 	return ok
 }
 
+// getSigners retrieves the signers of the given ECChain.
+func (q *quorumState) getSigners(value ECChain) bitfield.BitField {
+	head := value.HeadCIDOrZero()
+	chainSupport, ok := q.chainSupport[head]
+	signers := bitfield.New()
+	if !ok {
+		return signers
+	}
+
+	// Copy each element from the original map
+	for key := range chainSupport.signers {
+		signers.Set(uint64(q.powerTable.Lookup[key]))
+	}
+
+	return signers
+}
+
+// getSignatures returns the corresponding signatures for a given bitset of signers
+func (q *quorumState) getSignatures(value ECChain, signers bitfield.BitField) [][]byte {
+	head := value.HeadCIDOrZero()
+	signatures := make([][]byte, 0)
+	if err := signers.ForEach(func(i uint64) error {
+		if signature, ok := q.received[q.powerTable.Entries[i].ID].heads[head]; ok {
+			if len(signature) == 0 {
+				panic("signature is 0")
+			}
+			signatures = append(signatures, signature)
+		} else {
+			panic("QuorumSignature not found")
```

Comment:
Please don't introduce new panics, let's use error propagation

Reply:
Separate PR it is then.